portal-api/test-helpers/api-test-helpers/database/user-repository.js

```diff
@@ -0,0 +1,87 @@
+
+function issueJWTWithExpiryAndPayload({ user, sessionIdentifier, expiresIn, additionalPayload }) {
+  const { _id } = user;
+
+  const payload = {
+    sub: _id,
+    sessionIdentifier,
+    ...additionalPayload,
+  };
+
+  const signedToken = jsonwebtoken.sign(payload, PRIV_KEY, { expiresIn, algorithm: 'RS256' });
+
+  return {
+    token: `Bearer ${signedToken}`,
+    expires: expiresIn,
+    sessionIdentifier,
+  };
+}
+
+function issueValidUsernameAndPasswordJWT(user, sessionIdentifier) {
+  return issueJWTWithExpiryAndPayload({
+    user,
+    sessionIdentifier,
+    expiresIn: '105m',
+    additionalPayload: { username: user.username, loginStatus: LOGIN_STATUSES.VALID_USERNAME_AND_PASSWORD },
+  });
+}
+
+function issueValid2faJWT(user, sessionIdentifier) {
+  return issueJWTWithExpiryAndPayload({
+    user,
+    sessionIdentifier,
+    expiresIn: '12h',
+    additionalPayload: { username: user.username, roles: user.roles, bank: user.bank, loginStatus: LOGIN_STATUSES.VALID_2FA },
+  });
+}
+
+const overrideUserSignInTokenByUsername = async ({ username, newSignInToken }) => {
+  const salt = crypto.randomBytes(64);
+  const hash = crypto.pbkdf2Sync(newSignInToken, salt, 210000, 64, 'sha512');
+  const saltHex = salt.toString('hex');
+  const hashHex = hash.toString('hex');
+  const userCollection = await db.getCollection('users');
+  await userCollection.updateOne({ username: { $eq: username } }, { $set: { signInToken: { hashHex, saltHex } } });
+};
+
+// We can call this function on a user by user basis to create a partially logged in user session
+const createPartiallyLoggedInUserSession = async (user) => {
+  try {
+    const userCollection = await db.getCollection('users');
+    const userFromDatabase = await userCollection.findOne({ username: { $eq: user.username } }, { collation: { locale: 'en', strength: 2 } });
+
+    const sessionIdentifier = crypto.randomBytes(32).toString('hex');
+    const { token } = issueValidUsernameAndPasswordJWT(userFromDatabase, sessionIdentifier);
+    await userCollection.updateOne({ _id: { $eq: userFromDatabase._id } }, { $set: { sessionIdentifier } });
+
+    const signInToken = crypto.randomBytes(32).toString('hex');
+    await overrideUserSignInTokenByUsername({ username: user.username, newSignInToken: signInToken });
+
+    return { userId: userFromDatabase._id.toString(), token, signInToken }; // TODO DTFS2-6757: this is rather odd that we need to call to string
+  } catch (e) {
+    throw new Error(`Failed to create logged in user session for user: ${user.username}: ${e}`);
+  }
+};
+
+const createLoggedInUserSession = async (user) => {
+  try {
+    const userCollection = await db.getCollection('users');
+    const userFromDatabase = await userCollection.findOne({ username: { $eq: user.username } }, { collation: { locale: 'en', strength: 2 } });
+
+    const sessionIdentifier = crypto.randomBytes(32).toString('hex');
```

Comment:
Can we commonize these lines in this function and the previous one into a function like `getUserAndSessionIdFromDb`, for instance?

Reply:
I'm deliberately being a bit more expressive of whats going on here (as it's for tests rather than code) -- also they have enough variation between lines imo to justify not being commonised. What do you think?